pychromecast/controllers/spotify.py

```diff
@@ -0,0 +1,169 @@
+
+        data = {"remember": False, "username": username, "password": password,
+                "csrf_token": csrf_token}
+
+        response = session.post("https://accounts.spotify.com/api/login",
+                                data=data, cookies=cookies, headers=headers)
+
+        response.raise_for_status()
+
+    def _get_access_token(self, session, cookies):
+        """ Gets access token after login has been successful. """
+        headers = {'user-agent': USER_AGENT}
+
+        response = session.get("https://open.spotify.com/browse",
+                               headers=headers, cookies=cookies)
+        response.raise_for_status()
+
+        self.token = response.cookies['wp_access_token']
+
+        expiration = response.cookies['wp_expiration']
+        self.expiration_date = int(expiration)//1000
+
+    # Access token provided by spotipy does not have enough permissions to start
+    # Spotify on Chromecast
+    def start_session(self):
+        """ Starts session to get access token. """
+
+        # arbitrary value and can be static
+        cookies = {"__bon": "MHwwfC01ODc4MjExMzJ8LTI0Njg4NDg3NTQ0fDF8MXwxfDE="}
+
+        if self.username is None:
+            self.username = os.getenv("SPOTIFY_USERNAME")
+
+        if self.password is None:
+            self.password = os.getenv("SPOTIFY_PASS")
+
+        if self.username is None or self.password is None:
+            raise Exception("No username or password")
+
+        session = requests.Session()
+        token = self. _get_csrf(session, cookies)
+
+        self._login(session, cookies, self.username, self.password, token)
+        self._get_access_token(session, cookies)
+
+        self.session_started = True
+
+        return self.token
+
+    def launch_app(self):
+        """ Launch main application """
+
+        def callback():
+            """Callback function"""
+            self.send_message({"type": TYPE_STATUS, "credentials": self.token})
+
+        curr_time = time.time()
+
+        if self.session_started and curr_time < self.expiration_date:
+            self.logger.debug("Using same token: %s", self.token)
+            self.launch(callback_function=callback)
+        else:
+            self.logger.debug("Creating new token")
+            self.start_session()
+            self.logger.debug("Token is: %s", self.token)
+            self.launch(callback_function=callback)
+
+        # Need to wait for Spotify to be launched on Chromecast completely
+        time.sleep(5)
+        self.client = spotipy.Spotify(auth=self.token)
+
+        self.device_id = self.get_spotify_device_id()
+
+    def get_spotify_device_id(self):
+        """
+        Gets device id from Spotify. This should be the same as the name of
+        the Chromecast.
+        """
+        devices_available = self.client.devices()
+
+        for device in devices_available['devices']:
+            self.logger.debug(device)
+            if device['name'] == self.cast_name:
```

Comment:
Does `spotipy`'s `devices()` method return something like `UUID`? Because (almost) all chromecasts have a UUID and I think this code would be more robust if we wouldn't solely check the name.

Reply:
There is an id field which is the one that is being used to send the music to but afaik these can change at any moment.
The only things tying the device on the Spotify API and the Chromecast is name and perhaps the type which is set to `CastVideo`. Would checking for type to be equal to `CastVideo` make this a bit better?